File: Payload_Type/sage/container/commands/query.go
package commands

import (
	// Standard
	"fmt"
	"strings"

	// Internal
	"github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/anthropic"
	"github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/env"
	"github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/message"
	"github.com/MythicAgents/sage/Payload_Type/sage/container/pkg/openai"

	// Mythic
	structs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"
	"github.com/MythicMeta/MythicContainer/mythicrpc"
)

func query() structs.Command {
	attr := structs.CommandAttribute{
		SupportedOS: []string{"sage"},
	}

	provider := structs.CommandParameter{
		Name:                                    "provider",
		ModalDisplayName:                        "Provider",
		CLIName:                                 "provider",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_CHOOSE_ONE,
		Description:                             "The model provider to interact with (e.g. Anthropic, Bedrock, OpenAI)",
		Choices:                                 append([]string{""}, env.ProvidersString()...),
		DefaultValue:                            "",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       0,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       0,
				AdditionalInformation: nil,
			},
		},
	}

	model := structs.CommandParameter{
		Name:                                    "model",
		ModalDisplayName:                        "model",
		CLIName:                                 "model",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:                            "",
		Description:                             "The model to use for inference from the selected provider",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
		},
	}

	prompt := structs.CommandParameter{
		Name:                                    "prompt",
		ModalDisplayName:                        "prompt",
		CLIName:                                 "prompt",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:                            "",
		Description:                             "The prompt to send to the model for inference",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       2,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       2,
				AdditionalInformation: nil,
			},
		},
	}

	tools := structs.CommandParameter{
		Name:             "tools",
		ModalDisplayName: "tools",
		CLIName:          "tools",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_BOOLEAN,
		DefaultValue:     true,
		Description:      "Use tools to enhance the model's capabilities",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       3,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       3,
				AdditionalInformation: nil,
			},
		},
	}

	verbose := structs.CommandParameter{
		Name:             "verbose",
		ModalDisplayName: "Verbose",
		CLIName:          "verobse",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_BOOLEAN,
		DefaultValue:     false,
		Description:      "Show verbose output of all User & AI messages",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       4,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       4,
				AdditionalInformation: nil,
			},
		},
	}

	apiEndpoint := structs.CommandParameter{
		Name:             "API_ENDPOINT",
		ModalDisplayName: "API Endpoint",
		CLIName:          "API-ENDPOINT",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The API endpoint to use for the selected provider",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       5,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       5,
				AdditionalInformation: nil,
			},
		},
	}

	apiKey := structs.CommandParameter{
		Name:             "API_KEY",
		ModalDisplayName: "API Key",
		CLIName:          "API-KEY",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The API key to use for the selected provider",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       6,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       6,
				AdditionalInformation: nil,
			},
		},
	}

	awsAccessKey := structs.CommandParameter{
		Name:             "AWS_ACCESS_KEY_ID",
		ModalDisplayName: "AWS Access Key ID",
		CLIName:          "AWS-ACCESS-KEY-ID",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The AWS Access Key ID (AWS_ACCESS_KEY_ID) to use for Bedrock",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       7,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       7,
				AdditionalInformation: nil,
			},
		},
	}

	awsSecretAccessKey := structs.CommandParameter{
		Name:             "AWS_SECRET_ACCESS_KEY",
		ModalDisplayName: "AWS Secret Access Key",
		CLIName:          "AWS-SECRET-ACCESS-KEY",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The AWS Secret Access Key (AWS_SECRET_ACCESS_KEY) to use for Bedrock",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       8,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       8,
				AdditionalInformation: nil,
			},
		},
	}
	awsSessionToken := structs.CommandParameter{
		Name:             "AWS_SESSION_TOKEN",
		ModalDisplayName: "AWS Session Token",
		CLIName:          "AWS-SESSION-TOKEN",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The AWS Session Token (AWS_SESSION_TOKEN) to use for Bedrock",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       9,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       9,
				AdditionalInformation: nil,
			},
		},
	}
	awsRegion := structs.CommandParameter{
		Name:             "AWS_DEFAULT_REGION",
		ModalDisplayName: "AWS Default Region",
		CLIName:          "AWS-DEFAULT-REGION",
		ParameterType:    structs.COMMAND_PARAMETER_TYPE_STRING,
		DefaultValue:     "",
		Description:      "[OPTIONAL] The AWS Region (AWS_DEFAULT_REGION) to use for Bedrock",
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       10,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       10,
				AdditionalInformation: nil,
			},
		},
	}

	filename := structs.CommandParameter{
		Name:                                    "filename",
		ModalDisplayName:                        "Filename",
		CLIName:                                 "filename",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_CHOOSE_ONE,
		Description:                             "The filename to send to the model for inference",
		Choices:                                 []string{""},
		DefaultValue:                            "",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       11,
				AdditionalInformation: nil,
			},
		},
	}

	file := structs.CommandParameter{
		Name:                                    "file",
		ModalDisplayName:                        "File",
		CLIName:                                 "file",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_FILE,
		DefaultValue:                            "",
		Description:                             "The file to send to the model for inference",
		Choices:                                 nil,
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "New File",
				UIModalPosition:       11,
				AdditionalInformation: nil,
			},
		},
	}

	command := structs.Command{
		Name:                           "query",
		NeedsAdminPermissions:          false,
		HelpString:                     "query -prompt \"<prompt>\"",
		Description:                    "Send a single query to a model and get a single response",
		Version:                        0,
		SupportedUIFeatures:            nil,
		Author:                         "@Ne0nd0g",
		MitreAttackMappings:            []string{},
		ScriptOnlyCommand:              false,
		CommandAttributes:              attr,
		CommandParameters:              []structs.CommandParameter{provider, model, prompt, tools, verbose, apiEndpoint, apiKey, awsAccessKey, awsSecretAccessKey, awsSessionToken, awsRegion, file, filename},
		AssociatedBrowserScript:        nil,
		TaskFunctionOPSECPre:           nil,
		TaskFunctionCreateTasking:      queryCreateTask,
		TaskFunctionProcessResponse:    nil,
		TaskFunctionOPSECPost:          nil,
		TaskFunctionParseArgString:     taskFunctionParseArgString,
		TaskFunctionParseArgDictionary: taskFunctionParseArgDictionary,
		TaskCompletionFunctions:        nil,
	}

	return command
}

func queryCreateTask(task *structs.PTTaskMessageAllData) (resp structs.PTTaskCreateTaskingMessageResponse) {
	pkg := "mythic/container/commands/query/queryCreateTask()"
	resp.TaskID = task.Task.ID

	//data, filename, err := GetFile(task)
	//if err != nil {
	//	logging.LogError(err, "there was an error getting the file")
	//}
	//logging.LogDebug(fmt.Sprintf("Filename: %s", filename))

	provider, err := env.Get(task, "provider")
	if err != nil {
		err = fmt.Errorf("there was an error getting the 'provider' argument: %s", err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	model, err := env.Get(task, "model")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'model' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	prompt, err := task.Args.GetStringArg("prompt")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'prompt' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	tools, err := task.Args.GetBooleanArg("tools")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'tools' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	verbose, err := task.Args.GetBooleanArg("verbose")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'verbose' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	var output []message.Message

	respMsg := mythicrpc.MythicRPCResponseCreateMessage{
		TaskID:   task.Task.ID,
		Response: []byte(fmt.Sprintf("👤> %s\n", prompt)),
	}

	_, err = mythicrpc.SendMythicRPCResponseCreate(respMsg)
	if err != nil {
		resp.Error = fmt.Sprintf("Failed to send response: %s", err.Error())
		resp.Success = false
		logging.LogError(err, pkg)
		return
	}

	msg := message.Message{
		Role:    message.User,
		Content: prompt,
	}
	switch strings.ToLower(provider) {
	case "anthropic":
		output, err = anthropic.Chat(task, []message.Message{msg}, tools, verbose)
		if err != nil {
			resp.Error = fmt.Sprintf("Failed to invoke model: %s", err.Error())
			resp.Success = false
			logging.LogError(err, pkg)
			return
		}
	case "bedrock":
		if !strings.Contains(model, ".anthropic.") {
			err = fmt.Errorf("model '%s' is not supported by Bedrock", model)
			resp.Error = err.Error()
			resp.Success = false
			logging.LogError(err, pkg)
			return
		}
		output, err = anthropic.Chat(task, []message.Message{msg}, tools, verbose)
		if err != nil {
			resp.Error = fmt.Sprintf("Failed to invoke model: %s", err.Error())
			resp.Success = false
			logging.LogError(err, pkg)
			return
		}
	case "openai":
		output, err = openai.Chat(task, []message.Message{msg}, tools, verbose)
		if err != nil {
			resp.Error = fmt.Sprintf("Failed to invoke model: %s", err.Error())
			resp.Success = false
			logging.LogError(err, pkg)
			return
		}
	default:
		resp.Error = fmt.Sprintf("Unknown provider: %s", provider)
		resp.Success = false
		logging.LogError(fmt.Errorf("unknown provider: %s", provider), pkg)
		return
	}

	// Add the user's prompt to the output
	// Store the assistant message in the session and send the response to the user
	for k, o := range output {
		m := message.Message{
			Role:    message.Assistant,
			Content: o.Content,
		}
		sessions.UpdateMessages(resp.TaskID, m)

		if verbose {
			x := fmt.Sprintf("🤖> %s\n", o.Content)
			msg := mythicrpc.MythicRPCResponseCreateMessage{
				TaskID:   resp.TaskID,
				Response: []byte(x),
			}

			r, err := mythicrpc.SendMythicRPCResponseCreate(msg)
			if err != nil {
				resp.Error = fmt.Sprintf("Failed to send response: %s", err.Error())
				resp.Success = r.Success
				logging.LogError(err, pkg)
				return
			}
		} else if k == len(output)-1 {
			msg := mythicrpc.MythicRPCResponseCreateMessage{
				TaskID:   resp.TaskID,
				Response: []byte(fmt.Sprintf("🤖> %s\n", o.Content)),
			}

			r, err := mythicrpc.SendMythicRPCResponseCreate(msg)
			if err != nil {
				resp.Error = fmt.Sprintf("Failed to send response: %s", err.Error())
				resp.Success = r.Success
				logging.LogError(err, pkg)
				return
			}
		}
	}

	disp := fmt.Sprintf("with %s:%s", provider, model)
	resp.DisplayParams = &disp

	resp.Success = true
	resp.Completed = &resp.Success

	return
}

File: Payload_Type/sage/container/agent_functions/query.py
from mythic_container.MythicCommandBase import TaskArguments, CommandBase, CommandParameter, ParameterType, ParameterGroupInfo, PTTaskMessageAllData, PTTaskCreateTaskingMessageResponse
from mythic_container.MythicRPC import MythicRPCResponseCreateMessage, SendMythicRPCResponseCreate, MythicRPCCallbackUpdateMessage, SendMythicRPCCallbackUpdate, SendMythicRPCTaskUpdate, MythicRPCTaskUpdateMessage 
from mythic_container.logging import logger
from .utils import get_secret
from ai.langgraph.model import Model

class QueryArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)

        # Provider
        provider = CommandParameter(
            name="provider",
            display_name="Provider",
            cli_name="provider",
            type=ParameterType.String,
            description="The model provider to interact with (e.g. Anthropic, Bedrock, OpenAI)",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=3)]
        )

        # Model
        model = CommandParameter(
            name="model",
            display_name="Model",
            cli_name="model",
            type=ParameterType.String,
            description="The model to use for inference from the selected provider",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=4)]
        )
        # Prompt
        prompt = CommandParameter(
            name="prompt",
            display_name="Prompt",
            cli_name="prompt",
            type=ParameterType.String,
            description="The prompt to send to the model for inference",
            parameter_group_info=[ParameterGroupInfo(required=True, ui_position=0)]
        )
        # Tools
        tools = CommandParameter(
            name="tools",
            display_name="Tools",
            cli_name="tools",
            type=ParameterType.Boolean,
            default_value=True,
            description="Use tools to enhance the model's capabilities",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=1)]
        )
        # Verbose
        verbose = CommandParameter(
            name="verbose",
            display_name="Verbose",
            cli_name="verbose",
            type=ParameterType.Boolean,
            default_value=False,
            description="Show verbose output of all User & AI messages",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=2)]
        )
        # API Endpoint
        api_endpoint = CommandParameter(
            name="API_ENDPOINT",
            display_name="API Endpoint",
            cli_name="API_ENDPOINT",
            type=ParameterType.String,
            description="[OPTIONAL] The API endpoint to use for the selected provider",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=5)]
        )
        # API Key
        api_key = CommandParameter(
            name="API_KEY",
            display_name="API Key",
            cli_name="API_KEY",
            type=ParameterType.String,
            description="[OPTIONAL] The API key to use for the selected provider",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=6)]
        )
        # AWS Access Key ID
        aws_access_key = CommandParameter(
            name="AWS_ACCESS_KEY_ID",
            display_name="AWS_ACCESS_KEY_ID",
            cli_name="AWS_ACCESS_KEY_ID",
            type=ParameterType.String,
            description="[OPTIONAL] The AWS Access Key ID (AWS_ACCESS_KEY_ID) to use for Bedrock",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=7)]
        )
        # AWS Secret Access Key
        aws_secret_access_key = CommandParameter(
            name="AWS_SECRET_ACCESS_KEY",
            display_name="AWS_SECRET_ACCESS_KEY",
            cli_name="AWS_SECRET_ACCESS_KEY",
            type=ParameterType.String,
            description="[OPTIONAL] The AWS Secret Access Key (AWS_SECRET_ACCESS_KEY) to use for Bedrock",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=8)]
        )
        # AWS Session Token
        aws_session_token = CommandParameter(
            name="AWS_SESSION_TOKEN",
            display_name="AWS_SESSION_TOKEN",
            cli_name="AWS_SESSION_TOKEN",
            type=ParameterType.String,
            description="[OPTIONAL] The AWS Session Token (AWS_SESSION_TOKEN) to use for Bedrock",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=9)]
        )
        # AWS Region
        aws_region = CommandParameter(
            name="AWS_DEFAULT_REGION",
            display_name="AWS_DEFAULT_REGION",
            cli_name="AWS_DEFAULT_REGION",
            type=ParameterType.String,
            description="[OPTIONAL] The AWS Region (AWS_DEFAULT_REGION) to use for Bedrock",
            parameter_group_info=[ParameterGroupInfo(required=False,ui_position=10)]
        )

        # Add all the parameters
        self.args = [provider, model, prompt, tools, verbose, api_endpoint, api_key, aws_access_key, aws_secret_access_key, aws_session_token, aws_region]

    async def parse_arguments(self):
        if len(self.command_line) == 0:
            raise ValueError("Need to specify a prompt")
        self.add_arg("prompt", self.command_line)

    async def parse_dictionary(self, dictionary_arguments):
        if "prompt" not in dictionary_arguments:
            raise ValueError("Missing 'prompt' argument")
        for arg in dictionary_arguments:
            self.add_arg(arg, dictionary_arguments[arg])

class QueryCommand(CommandBase):
    cmd = "query"
    needs_admin = False
    help_cmd = "query -prompt <prompt>"
    description = "Send a single query to a model and get a single response"
    version = 1
    author = "@Ne0nd0g"
    argument_class = QueryArguments

    async def create_go_tasking(self, taskData: PTTaskMessageAllData) -> PTTaskCreateTaskingMessageResponse:
        response = PTTaskCreateTaskingMessageResponse(
            TaskID=taskData.Task.ID,
            Success=True,
        )

        aws_access_key_id: str | None = None
        aws_secret_access_key: str | None = None
        aws_session_token: str | None = None
        aws_region: str | None = None

        provider = get_secret(taskData=taskData, key="provider")
        if provider is None:
            response.Success = False
            response.Error = "unable to find the 'provider' key in the task, user secrets, payload build parameters, or in the payload container's environment variables"
            return response

        model = get_secret(taskData=taskData, key="model")
        if model is None:
            response.Success = False
            response.Error = "unable to find the 'model' key in the task, user secrets, payload build parameters, or in the payload container's environment variables"
            return response
        
        # Verify the prompt argument was provided and is not None
        prompt = taskData.args.get_arg(key="prompt")
        if prompt is None or not prompt:
            response.Success = False
            response.Error = "the 'prompt' argument must be provided and must not be blank"
            return response
        
        tools = taskData.args.get_arg("tools")
        verbose = taskData.args.get_arg("verbose")

        # These can be None as they are optional
        api_endpoint = get_secret(taskData=taskData, key="API_ENDPOINT")
        api_key = get_secret(taskData=taskData, key="API_KEY")
        if provider.lower() == "bedrock":
            aws_access_key_id = get_secret(taskData=taskData, key="AWS_ACCESS_KEY_ID")
            aws_secret_access_key = get_secret(taskData=taskData, key="AWS_SECRET_ACCESS_KEY")
            aws_session_token = get_secret(taskData=taskData, key="AWS_SESSION_TOKEN")
            aws_region = get_secret(taskData=taskData, key="AWS_DEFAULT_REGION")
        sys = get_secret(taskData=taskData, key="system_prompt")
        system_prompt = sys if sys or sys is not None else ""

        response.DisplayParams = f'{provider} {model} {prompt}'
        config = {"configurable":{}}
        if api_key is not None:
            config["configurable"]["api_key"] = api_key
        if api_endpoint is not None:
            config["configurable"]["base_url"] = api_endpoint
        if provider.lower() == "bedrock":
            if aws_access_key_id is not None:
                config["configurable"]["aws_access_key_id"] = aws_access_key_id
            if aws_secret_access_key is not None:
                config["configurable"]["aws_secret_access_key"] = aws_secret_access_key
            if aws_session_token is not None:
                config["configurable"]["aws_session_token"] = aws_session_token
            if aws_region is not None:
                config["configurable"]["region"] = aws_region
        llm = Model(provider=provider.lower(), model=model.lower(), system_prompt=system_prompt, config=config, task_id=taskData.Task.ID, agent_task_id=taskData.Task.AgentTaskID)
        await llm.initialize()
        if verbose:
            llm.set_verbose(True)

        await SendMythicRPCTaskUpdate(MythicRPCTaskUpdateMessage(TaskID=taskData.Task.ID, UpdateStatus="LLM Processing..."))
        await llm.invoke(prompt)

        resp = await SendMythicRPCCallbackUpdate(
            MythicRPCCallbackUpdateMessage(
                TaskID=taskData.Task.ID,
                UpdateLastCheckinTime=True,
                UpdateLastCheckinTimeViaC2Profile=""
            )
        )
        if not resp.Success:
            logger.error(f"Failed to update callback for task {taskData.Task.ID}: {resp.Error}")
        response.Completed = True
        return response
